app/lib/service/youtube/backend.dart

```diff
@@ -121,54 +121,63 @@ class _PkgOfWeekVideoFetcher {
             );
-        videos.addAll(rs!.items!.map(
-          (i) {
-            try {
-              final videoId = i.contentDetails?.videoId;
-              if (videoId == null) {
-                return null;
-              }
-              final thumbnails = i.snippet?.thumbnails;
-              if (thumbnails == null) {
-                return null;
-              }
-              final thumbnail = thumbnails.high ??
-                  thumbnails.default_ ??
-                  thumbnails.maxres ??
-                  thumbnails.standard ??
-                  thumbnails.medium;
-              final thumbnailUrl = thumbnail?.url;
-              if (thumbnailUrl == null || thumbnailUrl.isEmpty) {
-                return null;
-              }
-              return PkgOfWeekVideo(
-                videoId: videoId,
-                title: i.snippet?.title ?? '',
-                description:
-                    (i.snippet?.description ?? '').trim().split('\n').first,
-                thumbnailUrl: thumbnailUrl,
-              );
-            } catch (e, st) {
-              // this item will be skipped, the rest of the list may be displayed
-              _logger.pubNoticeShout(
-                  'youtube', 'Processing Youtube PlaylistItem failed.', e, st);
+
+        // process playlist items
+        for (final i in rs!.items!) {
+          try {
+            final videoId = i.contentDetails?.videoId;
+            if (videoId == null || videoIds.contains(videoId)) {
+              continue;
+            }
+            final thumbnails = i.snippet?.thumbnails;
+            if (thumbnails == null) {
+              continue;
+            }
+            final thumbnail = thumbnails.high ??
+                thumbnails.default_ ??
+                thumbnails.maxres ??
+                thumbnails.standard ??
+                thumbnails.medium;
+            final thumbnailUrl = thumbnail?.url;
+            if (thumbnailUrl == null || thumbnailUrl.isEmpty) {
+              continue;
             }
-            return null;
-          },
-        ).nonNulls);
-        // next page
+            videoIds.add(videoId);
+            videos.add(PkgOfWeekVideo(
+              videoId: videoId,
+              title: i.snippet?.title ?? '',
+              description:
+                  (i.snippet?.description ?? '').trim().split('\n').first,
+              thumbnailUrl: thumbnailUrl,
+            ));
+          } catch (e, st) {
+            // this item will be skipped, the rest of the list may be displayed
+            _logger.pubNoticeShout(
+                'youtube', 'Processing Youtube PlaylistItem failed.', e, st);
+          }
+        }
+
+        pageTokensVisited.add(nextPageToken ?? '');
+        // advance to next page token
         nextPageToken = rs.nextPageToken;
-        check = nextPageToken != null && nextPageToken.isNotEmpty;
+        if (nextPageToken == null ||
+            pageTokensVisited.contains(nextPageToken)) {
```

Comment:
Can this really happen?

If so, please explain in a comment. I would expect the nextPageToken to not link us in a cycle.

Reply:
Removed. It may be a bit more defensive coding than needed.